File: internal/sdk/typescript/eyrie.ts
/** Eyrie TypeScript SDK — conversation DAG client. */

export interface PromptOptions {
  model?: string;
  system_prompt?: string;
  max_tokens?: number;
  stream?: boolean;
}

export interface Node {
  id: string;
  parent_id?: string;
  root_id?: string;
  sequence: number;
  node_type: string;
  content: string;
  model?: string;
  provider?: string;
  tokens_in?: number;
  tokens_out?: number;
  created_at: string;
  title?: string;
}

export interface PromptResponse {
  content: string;
  node_id: string;
}

export interface StreamEvent {
  type: "delta" | "done" | "error";
  content?: string;
  node_id?: string;
  error?: string;
}

export class EyrieClient {
  private baseURL: string;
  private headers: Record<string, string>;

  constructor(baseURL = "http://localhost:8080", apiKey?: string) {
    this.baseURL = baseURL.replace(/\/$/, "");
    this.headers = { "Content-Type": "application/json" };
    if (apiKey) {
      this.headers["Authorization"] = `Bearer ${apiKey}`;
    }
  }

  async prompt(message: string, opts?: PromptOptions): Promise<PromptResponse> {
    const body = { message, ...opts };
    const resp = await fetch(`${this.baseURL}/prompt`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(body),
    });
    if (!resp.ok) throw new Error(`eyrie: ${resp.status} ${await resp.text()}`);
    return resp.json();
  }

  async promptFrom(nodeId: string, message: string, opts?: PromptOptions): Promise<PromptResponse> {
    const body = { message, ...opts };
    const resp = await fetch(`${this.baseURL}/nodes/${nodeId}/prompt`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(body),
    });
    if (!resp.ok) throw new Error(`eyrie: ${resp.status} ${await resp.text()}`);
    return resp.json();
  }

  async *streamPrompt(message: string, opts?: PromptOptions): AsyncGenerator<StreamEvent> {
    const body = { message, stream: true, ...opts };
    const resp = await fetch(`${this.baseURL}/prompt`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(body),
    });
    if (!resp.ok) throw new Error(`eyrie: ${resp.status} ${await resp.text()}`);
    const reader = resp.body!.getReader();
    const decoder = new TextDecoder();
    let buffer = "";
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop()!;
      for (const line of lines) {
        if (line.startsWith("data: ")) {
          yield JSON.parse(line.slice(6));
        }
      }
    }
  }

  async listConversations(): Promise<Node[]> {
    const resp = await fetch(`${this.baseURL}/nodes`, { headers: this.headers });
    if (!resp.ok) throw new Error(`eyrie: ${resp.status}`);
    return resp.json();
  }

  async getNode(nodeId: string): Promise<Node> {
    const resp = await fetch(`${this.baseURL}/nodes/${nodeId}`, { headers: this.headers });
    if (!resp.ok) throw new Error(`eyrie: ${resp.status}`);
    return resp.json();
  }

  async getTree(nodeId: string): Promise<Node[]> {
    const resp = await fetch(`${this.baseURL}/nodes/${nodeId}/tree`, { headers: this.headers });
    if (!resp.ok) throw new Error(`eyrie: ${resp.status}`);
    return resp.json();
  }

  async deleteNode(nodeId: string): Promise<void> {
    const resp = await fetch(`${this.baseURL}/nodes/${nodeId}`, { method: "DELETE", headers: this.headers });
    if (!resp.ok) throw new Error(`eyrie: ${resp.status}`);
  }

  async createAlias(nodeId: string, alias: string): Promise<void> {
    const resp = await fetch(`${this.baseURL}/nodes/${nodeId}/aliases/${alias}`, { method: "PUT", headers: this.headers });
    if (!resp.ok) throw new Error(`eyrie: ${resp.status}`);
  }

  async deleteAlias(alias: string): Promise<void> {
    const resp = await fetch(`${this.baseURL}/aliases/${alias}`, { method: "DELETE", headers: this.headers });
    if (!resp.ok) throw new Error(`eyrie: ${resp.status}`);
  }

  async health(): Promise<{ status: string }> {
    const resp = await fetch(`${this.baseURL}/health`);
    if (!resp.ok) throw new Error(`eyrie: ${resp.status}`);
    return resp.json();
  }
}

File: internal/sdk/python/eyrie.py
"""Eyrie Python SDK — conversation DAG client."""

import json
import httpx
from typing import Generator, Optional


class EyrieClient:
    def __init__(self, base_url: str = "http://localhost:8080", api_key: Optional[str] = None):
        self.base_url = base_url.rstrip("/")
        self.headers = {"Content-Type": "application/json"}
        if api_key:
            self.headers["Authorization"] = f"Bearer {api_key}"

    def prompt(self, message: str, model: str = "", system_prompt: str = "", max_tokens: int = 0) -> dict:
        body = {"message": message}
        if model:
            body["model"] = model
        if system_prompt:
            body["system_prompt"] = system_prompt
        if max_tokens:
            body["max_tokens"] = max_tokens
        resp = httpx.post(f"{self.base_url}/prompt", json=body, headers=self.headers, timeout=120)
        resp.raise_for_status()
        return resp.json()

    def prompt_from(self, node_id: str, message: str, model: str = "") -> dict:
        body = {"message": message}
        if model:
            body["model"] = model
        resp = httpx.post(f"{self.base_url}/nodes/{node_id}/prompt", json=body, headers=self.headers, timeout=120)
        resp.raise_for_status()
        return resp.json()

    def stream_prompt(self, message: str, model: str = "") -> Generator[dict, None, None]:
        body = {"message": message, "stream": True}
        if model:
            body["model"] = model
        with httpx.stream("POST", f"{self.base_url}/prompt", json=body, headers=self.headers, timeout=120) as resp:
            resp.raise_for_status()
            for line in resp.iter_lines():
                if line.startswith("data: "):
                    yield json.loads(line[6:])

    def list_conversations(self) -> list:
        resp = httpx.get(f"{self.base_url}/nodes", headers=self.headers, timeout=30)
        resp.raise_for_status()
        return resp.json()

    def get_node(self, node_id: str) -> dict:
        resp = httpx.get(f"{self.base_url}/nodes/{node_id}", headers=self.headers, timeout=30)
        resp.raise_for_status()
        return resp.json()

    def get_tree(self, node_id: str) -> list:
        resp = httpx.get(f"{self.base_url}/nodes/{node_id}/tree", headers=self.headers, timeout=30)
        resp.raise_for_status()
        return resp.json()

    def delete_node(self, node_id: str) -> dict:
        resp = httpx.delete(f"{self.base_url}/nodes/{node_id}", headers=self.headers, timeout=30)
        resp.raise_for_status()
        return resp.json()

    def create_alias(self, node_id: str, alias: str) -> dict:
        resp = httpx.put(f"{self.base_url}/nodes/{node_id}/aliases/{alias}", headers=self.headers, timeout=30)
        resp.raise_for_status()
        return resp.json()

    def delete_alias(self, alias: str) -> dict:
        resp = httpx.delete(f"{self.base_url}/aliases/{alias}", headers=self.headers, timeout=30)
        resp.raise_for_status()
        return resp.json()

    def health(self) -> dict:
        resp = httpx.get(f"{self.base_url}/health", timeout=5)
        resp.raise_for_status()
        return resp.json()
